### PR TITLE
Contract viewer [old - ignore]

### PR DESCRIPTION
**Pull request by hananbeer**
This is my first commit for code review.

The goal is to add a Contract Viewer tab with better functionalities than etherscan, ideally independent of etherscan and open source.

See demo video here:
https://x.com/high_byte/status/1743294802022609172?s=20

Things I've added:

- ui components live in `/components/ContractViewer.ts`. I use shadcn components which live in the `/@` folder, `mui-tree-viewer` element, and `monaco editor`. I think most of the CR should focus here since I'm no react developer.

- api endpoint in `/pages/api/getContract.ts` to fetch etherscan contract information

- utilities to flatten solidity source code (with small caveat of not yet supporting named imports) in `/util/flatten.ts`

- utilities to parse etherscan info in `/util/EtherscanApi.ts` and `/util/EtherscanParser.ts` (in the end we want to support sourcify too)

- reworked the solidity compiler web worker - supports selecting compiler version & generally just nicer smaller code.

- I don't use solc's ast for output, at least not yet. for that I use `@solidity-parser/parser` package. (that one was really annoying since it wasn't working in the browser.. fixed that)

### CODE REVIEW

File: components/ContractViewer/index.tsx [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+import NoSSR from 'react-no-ssr'
+
+import { state } from './ContractState'
+
+import {
+  ResizableHandle,
+  ResizablePanel,
+  ResizablePanelGroup,
+} from '@/components/ui/resizable'
+import ContractTreeView from './ContractTreeView'
```
Comment (charisra) [inline components/ContractViewer/index.tsx:16]: Try moving this up, on line 10

File: babel/solcWorker2.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+function onCompileRequest(msg) {
+  let { version, stdJson } = msg.data
+  if (!version || !stdJson) {
+    self.postMessage({ error: 'no version or standard json specified' })
+    return
+  }
+
+  // console.log('loading compiler:', url)
```
Comment (charisra) [inline babel/solcWorker2.js:27]: Remove all such console.log comments across all changed files when you're ready for a final review & have finished developing locally.

File: @/components/ui/resizable.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+const ResizableHandle = ({
+  withHandle,
+  className,
+  ...props
+}: React.ComponentProps<typeof ResizablePrimitive.PanelResizeHandle> & {
+  withHandle?: boolean
+}) => (
+  <ResizablePrimitive.PanelResizeHandle
+    className={cn(
+      "relative flex w-px items-center justify-center bg-border after:absolute after:inset-y-0 after:left-1/2 after:w-1 after:-translate-x-1/2 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring focus-visible:ring-offset-1 data-[panel-group-direction=vertical]:h-px data-[panel-group-direction=vertical]:w-full data-[panel-group-direction=vertical]:after:left-0 data-[panel-group-direction=vertical]:after:h-1 data-[panel-group-direction=vertical]:after:w-full data-[panel-group-direction=vertical]:after:-translate-y-1/2 data-[panel-group-direction=vertical]:after:translate-x-0 [&[data-panel-group-direction=vertical]>div]:rotate-90",
```
Comment (charisra) [inline @/components/ui/resizable.tsx:30]: these classes look really complex in all these components. Are all those needed and used?

File: components/ChainSelector.tsx [hunk 4]
```diff
@@ -46,18 +47,19 @@ const ChainSelector = () => {
...
     [onForkChange],
   )
 
-  useEffect(() => {
-    if (defaultForkOption) {
-      handleForkChange(defaultForkOption)
-    }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [defaultForkOption])
+  // TODO: don't redirect (or at least don't override address query param)
```
Comment (charisra) [inline components/ChainSelector.tsx:56]: Is this commented out block needed?

File: components/ContractViewer/ContractCodeEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  function handleEditorDidMount(editor, monaco) {
+    if (!editor) {
+      return
+    }
+
+    setCodeEditor(editor)
+
+    editor.updateOptions({ readOnly: true })
+    // monaco.languages.registerHoverProvider('sol', {
```
Comment (charisra) [inline components/ContractViewer/ContractCodeEditor.tsx:50]: Also, this block. Is it needed for it to stay?

File: components/ContractViewer/ContractState.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+const astParser = parser
+
+export class ContractInfo {
+  codeAddress: string
+  contextAddress: string
+  mainContractPath: string
+  code: string
+  originalPathLenses: any[] = []
+  // ast: any
+  // definitions tree - for contracts, functions, storage, etc.
```
Comment (charisra) [inline components/ContractViewer/ContractState.ts:21]: Remove?

File: components/ContractViewer/ContractState.ts [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  etherscanInfo: EtherscanContractResponse
+
+  parseAst() {
+    // parse ast and process definitions tree
+    const ast = astParser.parse(this.code, {
+      loc: true,
+      range: true,
+      tolerant: true,
+    })
+    //this.ast = ast
```
Comment (charisra) [inline components/ContractViewer/ContractState.ts:33]: And this

File: components/ContractViewer/ContractTreeNode.tsx [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+      }
+
+      return '🔧 function ' + node.name
+    },
+    widget: (node, root) => {
+      if (node.isConstructor) {
+        return null
+      }
+
+      // TODO: support
```
Comment (charisra) [inline components/ContractViewer/ContractTreeNode.tsx:79]: In a month from now you and no one else will be able to tell what this comment is about. Put a bit more info so you can have context when you restart working on it.

File: components/ContractViewer/ContractTreeNode.tsx [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      if (node.isFallback || node.isReceiveEther) {
+        return null
+      }
+
+      const [weiValue, setWeiValue] = useState(0n)
+      const [retValue, setRetValue] = useState<string>('')
```
Comment (charisra) [inline components/ContractViewer/ContractTreeNode.tsx:85]: You can omit <string> here it’s inferred from the default value

File: components/ContractViewer/index.tsx [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+  const [status, setStatus] = useState('Loading...')
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [currentAddress, setCurrentAddress] = useState<string>('')
+  const [currentCode, setCurrentCode] = useState<string>('')
+  const [codePeekLocation, setCodePeekLocation] = useState<any>({})
+
+  // TODO: move compilation to ContractsState?
+  const onCompilationResult = (event: MessageEvent) => {
+    // console.log('onCompilationResult', event.data)
```
Comment (charisra) [inline components/ContractViewer/index.tsx:42]: Remove console.log comment. There is more commented out code too, can it go?

File: @/components/ui/accordion.tsx [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    className="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
+    {...props}
+  >
+    <div className={cn("pb-4 pt-0", className)}>{children}</div>
+  </AccordionPrimitive.Content>
+))
+
+AccordionContent.displayName = AccordionPrimitive.Content.displayName
+
+export { Accordion, AccordionItem, AccordionTrigger, AccordionContent }
```
[the remark this answers is not in the document]
Comment (hananbeer) [reply @/components/ui/accordion.tsx:58]: as I mentioned, all components under `/@` are shadcn, not mine. this isn't exactly a ui lib, it lives inside your code. see: https://ui.shadcn.com/

File: components/ChainSelector.tsx (hunk 4 above)
Comment (hananbeer) [reply components/ChainSelector.tsx:56]: my bad it shouldn't be part of the commit. it removed my query param so instead of fixing I commented it out. I will fix it before the merge.

File: components/ContractViewer/index.tsx (hunk 10 above)
Comment (hananbeer) [reply components/ContractViewer/index.tsx:42]: why do you despise comments with such passion? comments are great!
really helps the flow of thought.
it can occasionally be trimmed, but I think obsessing about it on per-commit basis is counterproductive.

File: components/ContractViewer/ContractTreeView.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import ChevronRightIcon from '@mui/icons-material/ChevronRight'
+import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
+import { TreeView } from '@mui/x-tree-view/TreeView'
+
+import { ContractInfo } from './ContractState'
+import ContractTreeNode from './ContractTreeNode'
+
+type ContractTreeViewProps = {
+  forest: any[]
```
Comment (mcmxcdev) [inline components/ContractViewer/ContractTreeView.tsx:9]: `any` disables type checks completely, which is bad practice. Maybe you can update this to have proper type check in place.

File: components/ContractViewer/ContractTreeView.tsx [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  onSelect: (node, root) => void
+}
+
+const ContractTreeView = ({ forest, onSelect }: ContractTreeViewProps) => {
+  // let expanded: string[] = []
+  // expanded = forest.map((t) => t.defTree.id as string)
```
Comment (mcmxcdev) [inline components/ContractViewer/ContractTreeView.tsx:15]: Charis mentioned this already but all commented out code should either be commented in again or removed before we can merge this.

File: next.config.js [hunk 14]
```diff
@@ -1,16 +1,40 @@
 /** @type {import('next').NextConfig} */
+require('webpack')
+require('path')
```
Comment (mcmxcdev) [inline next.config.js:3]: Why did you add these imports? I don't think they don't do anything because you dont use anything from its namespace.

File: next.config.js [hunk 15]
```diff
@@ -1,16 +1,40 @@
...
 
 // eslint-disable-next-line @typescript-eslint/no-var-requires
 const { withPlausibleProxy } = require('next-plausible')
 
 module.exports = withPlausibleProxy()({
-  reactStrictMode: true,
+  reactStrictMode: false,
```
Comment (mcmxcdev) [inline next.config.js:9]: Fine to change to `false` temporarily, but you will need to reenable this again when you finish up this PR.

File: next.config.js [hunk 16]
```diff
@@ -1,16 +1,40 @@
...
   serverRuntimeConfig: {
     APP_ROOT: __dirname,
   },
+  compiler: {
+    styledComponents: false,
+  },
+  typescript: {
+    ignoreBuildErrors: true,
+  },
```
Comment (mcmxcdev) [inline next.config.js:18]: This is bad practice, please remove again.

File: next.config.js [hunk 17]
```diff
@@ -1,16 +1,40 @@
...
+
+    config.module.rules.push({
+      // We tell webpack to append "module.exports = SolidityParser;" at the end of the file.
+      test: require.resolve('@solidity-parser/parser/dist/index.iife.js'),
+      loader: 'exports-loader',
+      options: {
+        type: 'commonjs',
+        exports: 'single SolidityParser',
+      },
+    })
```
Comment (mcmxcdev) [inline next.config.js:36]: Looking at the referenced issue (which is closed now), it seems it was fixed in latest release: https://github.com/solidity-parser/parser/blob/master/CHANGELOG.md

File: pages/about.tsx [hunk 18]
```diff
@@ -369,24 +368,24 @@ const AboutPage = () => {
         </p>
       </SectionWrapper>
 
-      <SectionWrapper header={<H3>Access Sets</H3>} anchorKey="accesssets">
+      <SectionWrapper header={<H3>Access List</H3>} anchorKey="access_list">
         <p className="pb-6">
-          Access sets are defined per external transaction, and not per call.
+          Access list is defined per external transaction, and not per call.
```
Comment (mcmxcdev) [inline pages/about.tsx:373]: Any reason why you updated wording in this file? Pretty out of scope of this PR.

File: util/compiler.ts [hunk 19]
```diff
@@ -2,8 +2,9 @@ import { BN } from 'ethereumjs-util'
 import { IInstruction, IReferenceItem } from 'types'
 
 // Version here: https://github.com/ethereum/solc-bin/blob/gh-pages/bin/list.txt
-export const compilerSemVer = 'v0.8.19'
-export const compilerVersion = `soljson-${compilerSemVer}+commit.7dd6d404`
+// export const compilerSemVer = 'v0.8.19'
+// export const compilerVersion = `soljson-${compilerSemVer}+commit.7dd6d404`
+export const compilerVersion = `soljson-v0.8.19+commit.7dd6d404+commit.7dd6d404`
```
Comment (mcmxcdev) [inline util/compiler.ts:7]: Any reason why you changed this? Looks identical to me

File: tsconfig.json [hunk 20]
```diff
@@ -7,9 +7,10 @@
       "dom.iterable",
       "ESNext"
     ],
+    "sourceMap": true,
     "allowJs": true,
     "skipLibCheck": true,
-    "strict": true,
+    "strict": false,
```
Comment (mcmxcdev) [inline tsconfig.json:13]: This needs to be changed to `true` again, we want to keep code quality standards high.

File: package.json [hunk 21]
```diff
@@ -26,30 +29,60 @@
     "@kunigi/string-compression": "1.0.2",
     "@mdx-js/loader": "^2.3.0",
     "@mdx-js/react": "^2.3.0",
+    "@monaco-editor/react": "^4.6.0",
```
Comment (mcmxcdev) [inline package.json:32]: You added like 20 new dependencies which is something that usually needs to be discussed and agreed on.

You are mixing different styling approaches like MUI, RadixUI, TailwindCSS in one project which will blow up the bundle size.

**Review by 2xic — COMMENTED** [review]
I see you have added code for handling the various Etherscan response formats in [util/EtherscanParser.ts](https://github.com/smlxl/evm.codes/blob/60abb2f598694326e021d2f1d4b25fd0b6394534/util/EtherscanParser.ts) (nice!), but I still have issues viewing contracts that are not in the single source file format 

example of contracts that fails to be loaded: 
- **0x000000000022d473030f116ddee9f6b43ac78ba3** (standard solc json response)
- **0xfd1539ee9ccad1ab5d9b4877097cfd8221815b47** (multifile json response)

btw - I tried to debug this error myself, but when I try to run a preview (`yarn install` and then `yarn dev`) of this branch locally I get a error saying there are `"5 Unhandled Runtime Error"` (works fine on master). Do you see these errors also ?

**Comment by hananbeer** [comment]
> I see you have added code for handling the various Etherscan response formats (nice!) in [util/EtherscanParser.ts](https://github.com/smlxl/evm.codes/blob/60abb2f598694326e021d2f1d4b25fd0b6394534/util/EtherscanParser.ts), but I still have issues viewing contracts that are not in the single source file format
> 
> example of contracts that fails to be loaded:
> 
> * **0x000000000022d473030f116ddee9f6b43ac78ba3** (standard solc json response)
> * **0xfd1539ee9ccad1ab5d9b4877097cfd8221815b47** (multifile json response)
> 
> btw - I tried to debug this error myself, but when I try to run a preview (`yarn install` and then `yarn dev`) of this branch locally I get a error saying there are `"5 Unhandled Runtime Error"` (works fine on master). Do you see these errors also ?

thanks that's a good catch. some contracts have remappings and etherscan sources are pre-remapping, so I committed a fix to remap sources when looking for them.

**Comment by charisra** [comment]
Superseded by https://github.com/smlxl/evm.codes/pull/318
